src/conn.rs

```diff
@@ -114,40 +121,43 @@ impl<'a, 'c> InProgress<'a, 'c> {
                inputs)
     }
 
-    fn transact(self, transaction: &str) -> Result<(InProgress<'a, 'c>, TxReport)> {
+    pub fn transact(self, transaction: &str) -> Result<InProgress<'a, 'c>> {
         let assertion_vector = edn::parse::value(transaction)?;
         let entities = mentat_tx_parser::Tx::parse(&assertion_vector)?;
         self.transact_entities(entities)
     }
 
-    fn rollback(self) -> Result<()> {
+    pub fn last_report(&self) -> Option<Rc<TxReport>> {
+        self.last_report.clone()
+    }
+
+    pub fn rollback(mut self) -> Result<()> {
+        self.last_report = None;
         self.transaction.rollback().map_err(|e| e.into())
     }
 
-    fn commit(self) -> Result<()> {
-        {
-            // The mutex is taken during this block.
-            let mut metadata = self.mutex.lock().unwrap();
+    pub fn commit(self) -> Result<Option<Rc<TxReport>>> {
+        // The mutex is taken during this entire method.
+        let mut metadata = self.mutex.lock().unwrap();
 
-            if self.generation != metadata.generation {
-                // Somebody else wrote!
-                // Retrying is tracked by https://github.com/mozilla/mentat/issues/357.
-                // This should not occur -- an attempt to take a competing IMMEDIATE transaction
-                // will fail with `SQLITE_BUSY`, causing this function to abort.
-                bail!("Lost the transact() race!");
-            }
+        if self.generation != metadata.generation {
+            // Somebody else wrote!
+            // Retrying is tracked by https://github.com/mozilla/mentat/issues/357.
+            // This should not occur -- an attempt to take a competing IMMEDIATE transaction
+            // will fail with `SQLITE_BUSY`, causing this function to abort.
+            bail!("Lost the transact() race!");
+        }
 
-            // Commit the SQLite transaction while we hold the mutex.
-            self.transaction.commit()?;
+        // Commit the SQLite transaction while we hold the mutex.
+        self.transaction.commit()?;
 
-            metadata.generation += 1;
-            metadata.partition_map = self.partition_map;
-            if self.schema != *(metadata.schema) {
-                metadata.schema = Arc::new(self.schema);
-            }
+        metadata.generation += 1;
+        metadata.partition_map = self.partition_map;
+        if self.schema != *(metadata.schema) {
+            metadata.schema = Arc::new(self.schema);
         }
 
-        Ok(())
+        Ok(self.last_report)
```

Comment:
This feels wrong -- I think we should give back complete information (a vector of reports), and in the case of `rollback`, an empty `vec![]`.

Reply:
OTOH, doing it this way leaves `None` for no writes, and otherwise behaves just like you had written a series of linear independent `transact`s. I see your point, but I'd rather leave that work until we need it.